applications/serial_lte_modem/src/slm_ppp.c

```diff
@@ -0,0 +1,309 @@
+		return ret;
+	}
+
+	if (!open_ppp_sockets()) {
+		close_ppp_sockets();
+		net_if_down(ppp_iface);
+		return -ENOTCONN;
+	}
+
+	LOG_INF("PPP started.");
+
+	k_thread_create(&ppp_data_passing_thread_id, ppp_data_passing_thread_stack,
+			K_THREAD_STACK_SIZEOF(ppp_data_passing_thread_stack),
+			ppp_data_passing_thread, NULL, NULL, NULL,
+			K_PRIO_COOP(10), 0, K_NO_WAIT);
+	k_thread_name_set(&ppp_data_passing_thread_id, "ppp_data_passing");
+
+	return 0;
+}
+
+/** @return Whether PPP is fully started. */
+static bool slm_ppp_is_started(void)
+{
+	return net_if_flag_is_set(ppp_iface, NET_IF_UP)
+	    && ppp_fds[ZEPHYR_FD_IDX] >= 0
+	    && ppp_fds[MODEM_FD_IDX] >= 0;
+}
+
+static void slm_ppp_stop(void)
+{
+	if (!slm_ppp_is_started()) {
+		return;
+	}
+	/* First bring the interface down so that slm_ppp_is_started()
+	 * returns false right away. This is to prevent trying to stop PPP at the
+	 * same time from multiple sources: the original one (e.g. "AT#XPPP=0")
+	 * and the data passing thread. The latter attempts to stop PPP when it receives
+	 * an error on some of the sockets, which happens when they are closed.
+	 */
+	const int ret = net_if_down(ppp_iface);
+
+	if (ret) {
+		LOG_WRN("Failed to bring PPP interface down (%d).", ret);
+	}
+
+	close_ppp_sockets();
+
+	k_thread_join(&ppp_data_passing_thread_id, K_SECONDS(1));
+
+	LOG_INF("PPP stopped.");
+}
+
+int slm_ppp_init(void)
+{
+	if (!device_is_ready(ppp_uart_dev)) {
+		return -EAGAIN;
+	}
+
+	ppp_iface_idx = net_if_get_by_name(CONFIG_NET_PPP_DRV_NAME);
+	ppp_iface = net_if_get_by_index(ppp_iface_idx);
+	if (!ppp_iface) {
+		LOG_ERR("PPP network interface not found (%d).", ppp_iface_idx);
+		return -ENODEV;
+	}
+	net_if_flag_set(ppp_iface, NET_IF_POINTOPOINT);
+
+	LOG_DBG("PPP initialized.");
+	return 0;
+}
+
+/* Handles AT#XPPP commands. */
+int handle_at_ppp(enum at_cmd_type cmd_type)
+{
+	int ret;
+	unsigned int op;
+	enum {
+		OP_STOP,
+		OP_RESTART,
+		OP_COUNT
+	};
+
+	if (cmd_type != AT_CMD_TYPE_SET_COMMAND
+	 || at_params_valid_count_get(&slm_at_param_list) != 2) {
+		return -EINVAL;
+	}
+
+	ret = at_params_unsigned_int_get(&slm_at_param_list, 1, &op);
+	if (ret) {
+		return ret;
+	} else if (op >= OP_COUNT) {
+		return -ENOTSUP;
+	}
+
+	slm_ppp_stop();
+	if (op == OP_RESTART) {
+		return slm_ppp_start();
+	}
+	assert(op == OP_STOP);
+	return 0;
+}
+
+static void ppp_data_passing_thread(void*, void*, void*)
+{
+	const size_t mtu = net_if_get_mtu(ppp_iface);
+	struct pollfd fds[PPP_FDS_COUNT];
+
+	for (size_t i = 0; i != ARRAY_SIZE(fds); ++i) {
+		fds[i].fd = ppp_fds[i];
+		fds[i].events = POLLIN;
+	}
+
+	while (true) {
+		const int poll_ret = poll(fds, ARRAY_SIZE(fds), -1);
+
+		if (poll_ret <= 0) {
+			LOG_ERR("Sockets polling failed (%d, %d).", poll_ret, errno);
+			slm_ppp_stop();
+			return;
+		}
+
+		for (size_t src = 0; src != ARRAY_SIZE(fds); ++src) {
+			const short revents = fds[src].revents;
+
+			if (!revents) {
+				continue;
+			}
+			if (!(revents & POLLIN)) {
```

Comment:
How does this work if we have both POLLIN and POLLERR? I think that it then does only latter part and misses the POLLERR. Does the next poll then catch the error?

Reply:
This is the only one that I am really interested in this point.

Reply:
The way I understand it is that they cannot happen at the same time.
From `man poll`:
> The bits returned in revents can include any of those specified in events, or one of the values POLLERR, POLLHUP, or POLLNVAL.

Reply:
At least with TCP, it is possible to receive data in connection closure.

Reply:
Here we are dealing with raw sockets, with one end being the modem (not the network) and the other end being the PPP interface. From testing those simply return error events when they are shut down, and I don't think any of them would ever send any data along.
I still updated the condition to take this possibility into account.

Reply:
My point was to handle the possible POLLIN event before handling the error events. Now the logic will miss the possible data from POLLIN, should it be possible to receive that with errors.

Reply:
Oh well. Then the original implementation most likely handled it the way you wanted. If there were any errors along with a `POLLIN`, then I'm pretty sure it would instantly be received another time on the next loop iteration.

Reply:
It will likely receive another error when calling poll. I still think that it would be better to switch the order that you process the data in here, to first check the POLLIN and doing the data processing part and then doing the error handling part. but I leave it up to you. Especially since this will be continued.

Reply:
I consider it to be pretty solid as it is now, so will leave it like that. Switching the order would require another level of indentation, which would require quite some lines to be further split into several lines.